### PR TITLE
Mixer client tidying

### PR DESCRIPTION
**Pull request by dtebbs**
In preparation for client-side support for other pairings.

- [X] Better factoring in MixerClient to handle configuration parameters
- [x] Contract refactoring to abstract out curve-specific, as well as snark-specific, operations
  - [x] Adjust the location of residual bits to minimize the amount of pairing-specific code
  - [x] Move majority of functionality to `MixerBase`
- [x] Parallel updates to specification (clearmatics/zeth-specifications#1, clearmatics/zeth-specifications#3)

### CODE REVIEW

File: zeth_contracts/contracts/MixerBase.sol [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        public pure
+        returns (uint256 vpub_in, uint256 vpub_out)
+    {
+        // vpub_out and vpub_in occupy the first and second public_value_bits
+        vpub_out = (residual_bits & public_value_mask) * public_unit_value_wei;
+        vpub_in = ((residual_bits >> public_value_bits) & public_value_mask)
+        * public_unit_value_wei;
+    }
+
+    // This function is used to reassemble hsig given the the primary_inputs To
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/MixerBase.sol:231]: `given the the primary_inputs To` -> `given the primary_inputs. To` (`the` x2 and add `.`)

File: zeth_contracts/contracts/MixerBase.sol [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+        // The h_sig residual bits are after the jsIn authentication tags and
+        // jsIn nullifier bits.
+        return extract_bytes32(
+            primary_inputs[1 + jsIn + jsOut],
+            primary_inputs[1 + jsOut + num_hash_digests],
+            2 * jsIn);
+    }
+
+    // This function is used to reassemble the nullifiers given the nullifier
+    // index [0, jsIn[ and the primary_inputs To do so, we extract the
```
Comment (AntoineRondelet) [inline zeth_contracts/contracts/MixerBase.sol:246]: `the primary_inputs To do so,` -> `the primary_inputs. To do so,`

File: client/zeth/core/contracts.py [hunk 3]
```diff
@@ -375,3 +157,51 @@ def send_contract_call(
...
+    Query the attached node for all events emitted by the given contract
+    instance, with the given name. Yields an iterator of event-specific objects
+    to be decoded by the caller.
+    """
+    contract_address = instance.address
+    event_abi = find_matching_event_abi(instance.abi, event_name=event_name)
+    batch_size = batch_size or SYNC_BLOCKS_PER_BATCH
+
+    while start_block <= end_block:
+        # Filters are *inclusive* wrt "toBlock", hence the -1 here, and + 1 to
```
Comment (AntoineRondelet) [inline client/zeth/core/contracts.py:196]: `+ 1` -> `+1`

File: libzeth/serialization/proto_utils.cpp [hunk 4]
```diff
@@ -4,9 +4,31 @@
...
+
 namespace libzeth
 {
 
+// Note, any attempt to call an implementation not given will result in a
+// linker error.
+
+template<> std::string pp_name<libff::alt_bn128_pp>()
+{
+    return std::string("alt-bn128");
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libzeth/serialization/proto_utils.cpp:19]: Interesting. Yes, constants would be helpful in these cases.

File: client/zeth/core/mixer_client.py [hunk 5]
```diff
@@ -16,56 +16,34 @@
 from zeth.core.merkle_tree import MerkleTree, compute_merkle_path
 from zeth.core.pairing import PairingParameters
 import zeth.core.signing as signing
+import zeth.core.protoutils as protoutils
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply client/zeth/core/mixer_client.py]: renamed to `proto_utils`

File: scripts/test_zeth_cli_common.sh [hunk 6]
```diff
@@ -53,6 +53,14 @@ function setup_user_local_key() {
         (zeth_helper eth-gen-address && zeth_helper eth-fund)
     ! [ -e zeth-address.priv ] && \
         (zeth gen-address)
+
+    # Ensure that we have been able to perform all of these operations without
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply scripts/test_zeth_cli_common.sh:57]: removed extra `without`

File: zeth_contracts/contracts/Groth16AltBN128Mixer.sol [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    function verify_zk_proof(
+        uint256[] memory proof,
+        uint256[num_inputs] memory inputs)
+        internal
+        returns (bool)
+    {
+        // Convert the statically sized primaryInputs to a dynamic array
+        // expected by the verifyer.
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply zeth_contracts/contracts/Groth16AltBN128Mixer.sol]: fixed.

File: zeth_contracts/contracts/MixerBase.sol (hunk 1 above)
Comment (dtebbs) [reply zeth_contracts/contracts/MixerBase.sol]: fixed. thanks.

File: zeth_contracts/contracts/MixerBase.sol (hunk 2 above)
Comment (dtebbs) [reply zeth_contracts/contracts/MixerBase.sol:246]: fixed. thanks.

File: client/zeth/core/contracts.py (hunk 3 above)
Comment (dtebbs) [reply client/zeth/core/contracts.py]: fixed

File: client/zeth/core/mixer_client.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AntoineRondelet) [reply client/zeth/core/mixer_client.py]: Thanks

**Review by AntoineRondelet — APPROVED** [review]
LGTM - thanks @dtebbs

**Pull request merged**